hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenAirdropHandler.java

```diff
@@ -0,0 +1,567 @@
+import com.hedera.hapi.node.token.TokenAirdropTransactionBody;
+import com.hedera.hapi.node.token.TokenAssociateTransactionBody;
+import com.hedera.hapi.node.transaction.TransactionBody;
+import com.hedera.node.app.service.token.ReadableAccountStore;
+import com.hedera.node.app.service.token.ReadableNftStore;
+import com.hedera.node.app.service.token.ReadableTokenRelationStore;
+import com.hedera.node.app.service.token.ReadableTokenStore;
+import com.hedera.node.app.service.token.impl.WritableAccountStore;
+import com.hedera.node.app.service.token.impl.WritableAirdropStore;
+import com.hedera.node.app.service.token.impl.handlers.transfer.CryptoTransferExecutor;
+import com.hedera.node.app.service.token.impl.handlers.transfer.TransferContextImpl;
+import com.hedera.node.app.service.token.impl.util.AirdropHandlerHelper;
+import com.hedera.node.app.service.token.impl.validators.CryptoTransferValidator;
+import com.hedera.node.app.service.token.records.CryptoTransferRecordBuilder;
+import com.hedera.node.app.service.token.records.TokenAirdropRecordBuilder;
+import com.hedera.node.app.spi.fees.FeeContext;
+import com.hedera.node.app.spi.fees.Fees;
+import com.hedera.node.app.spi.workflows.HandleContext;
+import com.hedera.node.app.spi.workflows.HandleException;
+import com.hedera.node.app.spi.workflows.PreCheckException;
+import com.hedera.node.app.spi.workflows.PreHandleContext;
+import com.hedera.node.app.spi.workflows.TransactionHandler;
+import com.hedera.node.config.data.TokensConfig;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+/**
+ * This class contains all workflow-related functionality regarding {@link
+ * HederaFunctionality#TOKEN_AIRDROP}.
+ */
+@Singleton
+public class TokenAirdropHandler implements TransactionHandler {
+
+    private final CryptoTransferValidator validator;
+
+    /**
+     * Default constructor for injection.
+     */
+    @Inject
+    public TokenAirdropHandler(@NonNull final CryptoTransferValidator validator) {
+        this.validator = validator;
+    }
+
+    @Override
+    public void preHandle(@NonNull final PreHandleContext context) throws PreCheckException {
+        requireNonNull(context);
+
+        final var op = context.body().tokenAirdropOrThrow();
+        final var accountStore = context.createStore(ReadableAccountStore.class);
+        final var tokenStore = context.createStore(ReadableTokenStore.class);
+
+        for (final var transfers : op.tokenTransfers()) {
+            final var tokenID = transfers.tokenOrThrow();
+            final var tokenMeta = tokenStore.getTokenMeta(tokenID);
+            validateTruePreCheck(tokenMeta != null, INVALID_TOKEN_ID);
+            checkFungibleTokenTransfers(tokenID, transfers.transfers(), context, accountStore);
```

Comment:
⚠️ We want to use `preHandle()` **exclusively** for computing signing requirements (`Key`'s and hollow `Account`'s), and do all other business logic such as these validations in `handle()`.

The reason is that we will eventually want to optimize by skipping `preHandle()` at consensus when we can guarantee that no entity linked to the transaction has been added, deleted, or had its key changed since the consensus time of the immutable state used for `preHandle()`.

If `preHandle()` also contains critical business logic, we can't do that.

Reply:
+1

Reply:
Moved the business logic to `handle()` and we are now creating cryptoTransfer body and using `cryptoTransfer.preHandle` here